### PR TITLE
380 expose

### PR DESCRIPTION
**Pull request by rigazilla**
This PR allows the user to configure an external service in the Infinispan CR.
Syntax is:
```
spec:
  expose:
    type: LoadBalancer | NodePort
    port: 32033
```
`type` must be `LoadBalancer` or `NodePort`.
`port` is optional, if omitted will be `11222` for LoadBalancer or assigned by the platform for NodePort
This solves #380 point 1. Point 2 is postponed to 2.0.x

### CODE REVIEW

**Comment by Crumby** [comment]
@rigazilla I wonder if the `port` should be rather `nodePort` for the transparency? Or do you plan add other expose type in the feature and reuse `port` for non `nodePort` ports?

**Comment by rigazilla** [comment]
> @rigazilla I wonder if the `port` should be rather `nodePort` for the transparency? Or do you plan add other expose type in the feature and reuse `port` for non `nodePort` ports?

@Crumby, I think that `nodePort` is fine, it's strictly related to services, but I don't think we're going to provide in the Infinispan CR a way to customize port for future exposition options (i.e. route or ingress).

I'm going to apply you hint, thanks!

**Comment by rigazilla** [comment]
@Crumby changed to `nodePort`

File: pkg/apis/infinispan/v1/infinispan_types.go [hunk 1]
```diff
@@ -71,14 +71,37 @@ type InfinispanSiteLocationSpec struct {
...
+	// every node, in addition to 'ClusterIP' type.
+	ExposeTypeNodePort ExposeType = "NodePort"
+
+	// ExposeTypeLoadBalancer means a service will be exposed via an
+	// external load balancer (if the cloud provider supports it), in addition
+	// to 'NodePort' type.
+	ExposeTypeLoadBalancer ExposeType = "LoadBalancer"
+
+	// ExposeTypeRoute means Infinispan cluster will be exposed via Route resource
+	ExposeTypeRoute ExposeType = "Route"
```
Comment (Crumby) [inline pkg/apis/infinispan/v1/infinispan_types.go:96]: This new option doesn't seem to be handled later in the code.

File: test/e2e/main_test.go [hunk 2]
```diff
@@ -468,24 +468,22 @@ func TestExternalService(t *testing.T) {
...
-		return corev1.ServiceTypeClusterIP
 	case "NodePort":
-		return corev1.ServiceTypeNodePort
+		return ispnv1.ExposeTypeNodePort
 	case "LoadBalancer":
-		return corev1.ServiceTypeLoadBalancer
-	case "ExternalName":
-		return corev1.ServiceTypeExternalName
+		return ispnv1.ExposeTypeLoadBalancer
+	case "ExternalName", "ClusterIP":
```
Comment (Crumby) [inline test/e2e/main_test.go:485]: Could thsi option be just dropped as it's behaviour is same as `default`?

File: test/e2e/main_test.go [hunk 3]
```diff
@@ -468,24 +468,22 @@ func TestExternalService(t *testing.T) {
...
+	return ispnv1.ExposeSpec{
+		Type:     exposeServiceType(),
+		NodePort: 30222,
 	}
 }
 
-func exposeServiceType() corev1.ServiceType {
+func exposeServiceType() ispnv1.ExposeType {
 	exposeServiceType := getEnvWithDefault("EXPOSE_SERVICE_TYPE", "NodePort")
 	switch exposeServiceType {
```
Comment (Crumby) [inline test/e2e/main_test.go:480]: Missing `Route` option?

File: pkg/controller/infinispan/infinispan_controller.go [hunk 4]
```diff
@@ -285,23 +285,26 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
...
-				if err != nil && !errors.IsAlreadyExists(err) {
-					reqLogger.Error(err, "failed to create external Service", "Service", ser)
-					return reconcile.Result{}, err
-				}
-				infinispan.Spec.Expose = externalService.Spec
-				err = r.client.Update(context.TODO(), infinispan)
-				if err != nil {
-					reqLogger.Error(err, "failed to Infinispan with Service spec", "Service", ser)
-					return reconcile.Result{}, err
+			switch infinispan.Spec.Expose.Type {
```
Comment (Crumby) [inline pkg/controller/infinispan/infinispan_controller.go:288]: Missing `ExposeTypeRoute` option?

File: pkg/apis/infinispan/v1/infinispan_types.go (hunk 1 above)
Comment (rigazilla) [reply pkg/apis/infinispan/v1/infinispan_types.go:96]: I've tried to add support for Route exposition, but openshift packages conflict with the current dependencies set, so I've posponed this feature to 2.x.
It's probably better to remove all the references to Route for now. Going to do.

File: test/e2e/main_test.go (hunk 3 above)
Comment (rigazilla) [reply test/e2e/main_test.go:480]: se above

File: pkg/controller/infinispan/infinispan_controller.go (hunk 4 above)
Comment (rigazilla) [reply pkg/controller/infinispan/infinispan_controller.go:288]: see above

File: test/e2e/main_test.go (hunk 2 above)
Comment (rigazilla) [reply test/e2e/main_test.go:485]: yeah, fine

**Comment by rigazilla** [comment]
@oraNod I've updated the docs with this changes

**Comment by rigazilla** [comment]
merged, thanks!

**Pull request merged**